pandas/_libs/tslibs/src/datetime/np_datetime_strings.c

```diff
@@ -104,19 +105,30 @@ int parse_iso_8601_datetime(const char *str, int len, int want_exc,
     while (sublen > 0 && isspace(*substr)) {
         ++substr;
         --sublen;
+        if (format_len < 1 && exact) goto parse_error;
+        if (format_len && *format++ != ' ') goto parse_error;
+        if (format_len) --format_len;
     }
 
     /* Leading '-' sign for negative year */
     if (*substr == '-') {
         ++substr;
         --sublen;
+        if (format_len < 1 && exact) goto parse_error;
+        if (format_len && *format++ != '-') goto parse_error;
+        if (format_len) --format_len;
     }
 
     if (sublen == 0) {
         goto parse_error;
     }
 
     /* PARSE THE YEAR (4 digits) */
+    if (format_len < 2 && exact) goto parse_error;
```

Comment:
```suggestion
    if (format_len < 2) {
      if (exact) {
        goto parse_error;
      }
      // What happens here?
    } else {
      if (*format++ != '%' && *format++ != 'Y') {
        goto parse_error;
      }
      
      format_len -= 2;
    }
```

Current code has undefined behavior and a possible segfault if `format_len` is 1 while `exact` is False. Need to think through what action should be taken there

Reply:
In practice that shouldn't be possible - this function's only used internally, and 

https://github.com/pandas-dev/pandas/blob/57d8d3a7cc2c4afc8746bf774b5062fa70c0f5fd/pandas/_libs/tslibs/parsing.pyx#L886-L903

would return false if the format were for example `%Y-%`. I've changed it to raise if `format_len` isn't `0` anyway, thanks

Reply:
Yea what you will find with C is that you want to be extremely explicit. Even if it's not possible today, if someone refactors the code and it does happen things become very difficult to debug. 

Our error checking in some of our extensions is really loose, which is why our internal code is relatively difficult to refactor